File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        sortNum = sorted(nums)
        for i in range(len(sortNum)):
            if i > 0 and sortNum[i] == sortNum[i - 1]:
                continue
            j = i + 1
            k = len(sortNum) - 1
            while(j < k):

                sum3 = sortNum[i] + sortNum[j] + sortNum[k]
                if sum3 > 0:
                    k -= 1
                elif sum3 < 0:
                    j += 1
                else:
                    res.append([sortNum[i], sortNum[j], sortNum[k]])
                    j += 1
                    k -= 1
                    while j < k and sortNum[j] == sortNum[j - 1]:
                        j += 1

                    while j < k and sortNum[k] == sortNum[k + 1]:
                        k -= 1

        return res

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int previousOffset;
        vector<vector<int>> res;
        for(int i = 0; i < nums.size(); i++)
        {
            if(previousOffset == nums[i]) continue;
            int low = i + 1;
            int high = nums.size() - 1;

            while(high > low)
            {
                int sum = nums[i] + nums[low] + nums[high];
                if(sum > 0) high--;
                else if (sum < 0) low++;
                else
                {
                    vector<int> toBeAdded = {nums[i], nums[low], nums[high]};
                    if(res.empty() || (!res.empty() && res.back() != toBeAdded)) res.push_back(toBeAdded);
                    low++;
                }
            }
            previousOffset = nums[i];
        }
        return res;
    }
};